sdk/core/core/inc/az_span.h

```diff
@@ -260,6 +260,32 @@ AZ_NODISCARD az_result az_span_to_uint64(az_span span, uint64_t* out_number);
  */
 AZ_NODISCARD az_result az_span_to_uint32(az_span span, uint32_t* out_number);
 
+/**
+ * @brief az_span_find searches for `target` in `source`, returning an #az_span within `source` if it finds it.
+ *
+ * @param[in] source The #az_span with the content to be searched on.
+ * @param[in] target The #az_span containing the token to be searched in `source`.
+ * @param[out] out_span The #az_span pointing to the first occurrence of `target` in `source`, if it is found.
+ * @return An #az_result value indicating the result of the operation.
+ *          #AZ_OK if `target` is found in `source`
+ *          #AZ_ERROR_ITEM_NOT_FOUND if `target` is not found in `source`
```

Comment:
Is it really an "error" condition to not find the `target` in the `source`? To me, not finding it is normal part of querying the source span. Maybe we can drop `ERROR_` and just have `AZ_ITEM_NOT_FOUND`.

We should probably also document that, if not found `out_span` will be `NULL`.

Reply:
By definition, an error is when the function doesn't say it will do (as indicated by the function name). So, if az_span_find fails to find, then it is an error - by definition.
It is also a convention that if an error is returned, then the customer should ignore the out parameter. We can set it to null but the "contract" is that the customer should not look at it. If we doc that we set it to null, this is a contract which is in opposition to our convention.

Reply:
By definition, an error is when a function fails to do what it says it will do (as indicated by the function name). So, if "find" fails to "find" then it is an error.

Also, our convention is that the customer should not look at out parameters if the function returns an error. Documenting that, on error, out_span will be null ties us to a contract that should not matter to the customer and is in violation of our convention. 

We could change our convention to always set output parameters to 0/null on error but it is odd to say this if customers should not look at it if an error is returned.

Reply:
I was looking at APIs like C++ string::find and it returns [npos](http://www.cplusplus.com/reference/string/string/npos/) when the target isn't found, rather than throwing an exception (similarly .NET's IndexOf returns -1 rather than throwing an error).
http://www.cplusplus.com/reference/string/string/find/

Since C doesn't have exceptions, I read "ERROR" containing az_result codes as capturing those exception conditions. For example, C++ string::push_back throws an exception, which if we were to implement here would be an error code:
http://www.cplusplus.com/reference/string/string/push_back/

Reply:
.NET doesn't follow its own guidelines and I can point you to many examples that do not follow the rules. IndexOf is just one example.  Stream's ReadByte is another example.
C/C++ examples are also not ideal. The problem is that the way of thinking has changed over the years and many examples exist which retain the old way of thinking for backward compatibility reasons. 

Similarly, lots of people say exceptions should only be used in exceptional circumstances or don't use exceptions for code flow but these also don't make sense. What is a clear definition for "exceptional circumstance" and, the whole point of exceptions is to alter code flow similar to goto, continue, break, return, and similar language constructs. Why is break or return OK to prematurely exit from code flow but an exception is not OK? 

Some people would say perf but this is language, compiler, runtime, and OS specific. Exceptions are not bad perf in many combinations of these. Furthermore, we should not focus on premature perf optimizations. So, the "modern" thinking is that an error is when a function can't do what it says it will do and, when this happens, the function reports error via an error or exception - this is an easy to follow rule. 

Also, contract violations do NOT fall into this category: a function is allowed to behave unpredictably in response to contract violations.